Clickhouse: Add support for [LEFT] ARRAY JOIN

simpl1g opened the pull request:
<!--Thanks for adding this feature!-->

<!--Please give the Pull Request a meaningful title for the release notes-->

### Brief summary of the change made
Clickhouse: Add support for [LEFT] ARRAY JOIN

### Are there any other side effects of this change that we should be aware of?


### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

konnectr commented:
It seems to me that you made broken visions.
```
SELECT * FROM (SELECT [1, 2] AS arr,[1, 2] AS arr2) AS t1 ARRAY JOIN arr,arr2;
```

konnectr commented:
There is another example
```
SELECT *,ch,cg
FROM (SELECT 1)
ARRAY JOIN 
  ['1','2','3','4','5'] as cg,
  splitByChar(',','1,2,3,4,5') as ch
```
https://fiddle.clickhouse.com/cd078126-fd2d-4222-b919-de0ffb1aeb55

simpl1g commented:
> It seems to me that you made broken visions.
> 
> ```
> SELECT * FROM (SELECT [1, 2] AS arr,[1, 2] AS arr2) AS t1 ARRAY JOIN arr,arr2;
> ```
> 
> arr2 gets into from_expression, but in theory it should also be in join_clause

@konnectr Thank you for review, I fixed your comments and added examples to test cases.
There is no `NATURAL JOIN` in ClickHouse, so I replaced it with `ARRAY JOIN` implementation

konnectr commented:
```
echo "SELECT *,ch,cg FROM (SELECT 1)                                                                                                                                                                                          
  ARRAY JOIN ['1','2'] as cg
  ARRAY JOIN splitByChar(',','1,2') as ch;" | sqlfluff parse --dialect clickhouse -
```
https://fiddle.clickhouse.com/24e57ea2-afe9-4e0e-ade8-c129621324cd

In theory it shouldn't work.

WittierDinosaur commented on src/sqlfluff/dialects/dialect_clickhouse.py:
```diff
@@ -177,13 +186,26 @@ class JoinClauseSegment(ansi.JoinClauseSegment):
         ),
-        # Note NATURAL joins do not support Join conditions
-        Sequence(
-            Ref("JoinKeywordsGrammar"),
```
Just to confirm, did you intend to delete this block?

simpl1g replied:
Yes, ClickHouse does not support `NATURAL JOIN` and this block was added here by mistake

WittierDinosaur reviewed (approved):
LGTM
